bearssl: support CURLOPT_CAINFO_BLOB

AzazKamaz opened the pull request:
Just implement support to use `CURLOPT_CAINFO_BLOB` with bearssl.
Also fixed bearssl support building error on `hostname = NULL;` while having `const char * const hostname = SSL_HOST_NAME();`

There are four ways to implement support:
- have one function with parsing cert file from the buffer and two ways to get buffer. I don't want to read an entire file into memory so denied
- have two big functions with similar code buf from different sources. Too big code repeat so denied
- parse cert file iteratively while having state struct. Need to split function into some struct, init(), dispose() and update(). Denied because of complexity
- have some switches in the cert parsing function. The easiest way with a minimum amount of modification. Implemented this way

danielgustafsson commented on lib/vtls/bearssl.c:
```diff
@@ -340,8 +365,29 @@ static CURLcode bearssl_connect_step1(struct Curl_easy *data,
+    struct cafile_source source = {
+      .type = CAFILE_SOURCE_BLOB,
+      .data = ca_info_blob->data,
+      .len = ca_info_blob->len,
```
I might be undecaffeinated, but isn't this C99 syntax?

AzazKamaz replied:
Right. Changed to the older way

danielgustafsson commented on lib/vtls/bearssl.c:
```diff
@@ -100,26 +109,38 @@ static CURLcode load_cafile(const char *path, br_x509_trust_anchor **anchors,
+  do {
+    if(source->type == CAFILE_SOURCE_PATH) {
+      n = fread(buf, 1, sizeof(buf), fp);
+      p = buf;
```
Assigning p to buf for an n of zero is a change from previous coding which doesn't seem right.

AzazKamaz replied:
Since buf is just a local array I think the only difference is a bit more instructions executed per function call. Fixed

danielgustafsson commented on lib/vtls/bearssl.c:
```diff
@@ -100,26 +109,38 @@ static CURLcode load_cafile(const char *path, br_x509_trust_anchor **anchors,
-  fp = fopen(path, "rb");
-  if(!fp)
+  if(source->type != CAFILE_SOURCE_PATH && source->type != CAFILE_SOURCE_BLOB)
     return CURLE_SSL_CACERT_BADFILE;
```
This doesn't seem like the best/right errcode, don't we have a better one to indicate incorrect usage?

AzazKamaz replied:
You are right, changed to `CURLE_BAD_FUNCTION_ARGUMENT`

jsoref commented on lib/vtls/bearssl.c:
```diff
@@ -340,8 +365,29 @@ static CURLcode bearssl_connect_step1(struct Curl_easy *data,
+        failf(data, "error importing CA certificate blob");
+        return ret;
+      }
+      /* Only warning if no certificate verification is required. */
```
did you mean:
```suggestion
      /* Only warn if no certificate verification is required. */
```

AzazKamaz replied:
I have just copied from the same openssl part :)

AzazKamaz commented:
Actually, I am force pushing my branch to stay away from commits like "fix formatting" or "change initialization way"

bagder commented on lib/vtls/bearssl.c:
```diff
@@ -100,26 +109,38 @@ static CURLcode load_cafile(const char *path, br_x509_trust_anchor **anchors,
-  if(!fp)
-    return CURLE_SSL_CACERT_BADFILE;
+  if(source->type != CAFILE_SOURCE_PATH && source->type != CAFILE_SOURCE_BLOB)
+    return CURLE_BAD_FUNCTION_ARGUMENT;
```
This can't happen can it the way the code is written and intended? Shouldn't this then rather be made with DEBUGASSERT() instead?

bagder reviewed (commented):
The `CURLOPT_CAINFO_BLOB.3` man page need an update to clarify that bearssl also supports this option from >= 7.79.0, right?

AzazKamaz commented:
I have added info to docs and used DEBUGASSERT for check.
Also I have changed `source->len ? source->len : strlen(source->data)` to `source->len` because documentation requires pointer **and** size and strlen will cause problems if buffer is zero-length and not zero-determinated.
And added check `source->len <= (size_t)INT_MAX` same as in OpenSSL backend

bagder commented:
Thanks!
